### PR TITLE
Budget command added allowing user to specify their budget for a specific month

### LINKED ISSUE #20: As a phone junkie, I can set aside enough money for purchasing the latest phones and accessories.

### LINKED ISSUE #24: As a student who likes to buy games online, I can view the monthly funds left
... to see if I am able to pay my school fees.

### LINKED ISSUE #26: As a mobile game addict who makes a lot of in-app purchases, I can check my funds remaining for the month
... to manage my finances and to sustain my essential expenses.

### LINKED ISSUE #30: As a frugal NUS student, I can view the funds left for the month
... so that I can optimise my finances.

### LINKED ISSUE #31: As a social animal frequently going to clubs and bars, I can view my monthly budget left and plan out my expenses
... so I can moderate my spending on social activities and keep up my balanced lifestyle.

### LINKED ISSUE #35: As a student who is currently in a relationship, I need the tracker to help me make sure I am not exceeding my monthly budget
... for both my partner and I, so that we can get married soon.

### LINKED ISSUE #84: Filter TransactionList on startup to only show the current month transactions, calculate the balance for the current month

### LINKED ISSUE #97: Set monthly budget limit and display current month in GUI
The current month's budget and balance will be shown, unless changed by a `list` command that specifies a different month, during which the month, budget and balance will be displayed instead.

### LINKED ISSUE #144: Modify add/update/delete to update budget and balance for the month of the affected transaction

### LINKED ISSUE #145: Store and read the BudgetList from/to disk

### PR DESCRIPTION
**Pull request by limerencee**
Month are uniquely identified by the following format: `MM/yyyy`.

Other changes:
1. Updated UGDG to include `budget` command.
2. Updated prefix for money from `c/COST` to `v/VALUE`, UGDG are updated to reflect this change.
3. Added classes to support `BudgetList` which stores `Budget` objects containing `Value` and `Calendar`.
4. Added JUnit test cases for `budget` command and the supporting classes like `BudgetList`, `BudgetCommand`, `BudgetCommandParser` and `ThriftTest` itself.
5. Setting budget now saves into the JSON file, similar to how `Transaction` is stored.
6. Added JUnit tests for (de)/serializing of `Budget` objects.
7. Fixed JUnit tests for `JsonSerializableThrift`.
8. Added `StreamUtils` for converting `Iterator<T>` to `Stream<T>`. JUnit tests are also added for this new utility class.
9. New `TransactionIsInMonthYearPredicateTest` file to check if a `Transaction` belongs to that month. JUnit tests are added as well.
10. Current month's budget is shown in GUI.
11. Current month's balance is shown in GUI.
12. After parsing successfully either a Add/Budget/Delete/Update command, reflect the possible changes on the GUI.

*Ignore the following tags:*
Resolves #20 Resolves #24 Resolves #26 Resolves #30 Resolves #31 Resolves #35 Resolves #84 Resolves #97 Resolves #144 Resolves #145

### CODE REVIEW

File: docs/DeveloperGuide.adoc [hunk 1]
```diff
@@ -53,7 +53,7 @@ The *_Architecture Diagram_* given above explains the high-level design of the A
 The `.puml` files used to create diagrams in this document can be found in the link:{repoURL}/docs/diagrams/[diagrams] folder.
 Refer to the <<UsingPlantUml#, Using PlantUML guide>> to learn how to create and edit diagrams.
 
-`Main` has two classes called link:{repoURL}/src/main/java/seedu/address/Main.java[`Main`] and link:{repoURL}/src/main/java/seedu/address/MainApp.java[`MainApp`]. It is responsible for,
+`Main` has two classes called link:{repoURL}/srv/main/java/seedu/address/Main.java[`Main`] and link:{repoURL}/srv/main/java/seedu/address/MainApp.java[`MainApp`]. It is responsible for,
```
Comment (lightz96) [inline docs/DeveloperGuide.adoc]: Why is it change from `src` to `srv`?

File: src/main/java/thrift/logic/Logic.java [hunk 2]
```diff
@@ -35,9 +38,24 @@ CommandResult execute(String commandText, TransactionListPanel transactionListPa
      */
     ReadOnlyThrift getThrift();
 
-    /** Returns an unmodifiable view of the filtered list of transactions*/
+    /** Returns if the given command requires a refresh of the filteredlist. */
+    boolean requireFilteredListRefresh(Command command);
```
Comment (lightz96) [inline src/main/java/thrift/logic/Logic.java]: Boolean methods should be named to sound like booleans.

File: src/main/java/thrift/logic/parser/BudgetCommandParser.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Parses input arguments and creates a new BudgetCommand object.
+ */
+public class BudgetCommandParser implements Parser<BudgetCommand> {
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the BudgetCommand and returns an BudgetCommand
+     * object for execution.
+     * @throws ParseException if the user input does not conform to the expected format.
```
Comment (lightz96) [inline src/main/java/thrift/logic/parser/BudgetCommandParser.java:20]: There should be an empty line between description and parameter section.

File: src/main/java/thrift/model/Model.java [hunk 4]
```diff
@@ -125,16 +131,35 @@
...
 
     /** Returns an unmodifiable view of the filtered transaction list */
     ObservableList<Transaction> getFilteredTransactionList();
 
+    /** Filters the view of the transaction list to only show transactions that occur in the current month. */
+    void updateFilteredTransactionListToCurrentMonth();
+
     /**
      * Updates the filter of the filtered transaction list to filter by the given {@code predicate}.
      * @throws NullPointerException if {@code predicate} is null.
```
Comment (lightz96) [inline src/main/java/thrift/model/Model.java:149]: There should be an empty line between description and parameter section

File: src/main/java/thrift/model/Model.java [hunk 5]
```diff
@@ -125,16 +131,35 @@
...
     void updateFilteredTransactionList(Predicate<Transaction> predicate);
 
+    /**
+     * Updates the balance tracked by the model by summing values from the {@code Transaction} in the filteredList.
+     */
+    void updateBalanceForCurrentMonth();
+
+    /**
+     * Returns the balance held by the model to update the GUI with.
+     * @return
```
Comment (lightz96) [inline src/main/java/thrift/model/Model.java]: Either remove `@return` or describe what is returned

File: src/main/java/thrift/model/transaction/TransactionIsInMonthYearPredicate.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof TransactionIsInMonthYearPredicate) // instanceof handles nulls
+                && (((TransactionIsInMonthYearPredicate) other).monthYear.get(Calendar.MONTH)
+                    == monthYear.get(Calendar.MONTH)
+                    && ((TransactionIsInMonthYearPredicate) other).monthYear.get(Calendar.YEAR)
+                    == monthYear.get(Calendar.YEAR));
```
Comment (lightz96) [inline src/main/java/thrift/model/transaction/TransactionIsInMonthYearPredicate.java]: It should be indented by 8 spaces. 🤔

File: src/main/java/thrift/model/transaction/Value.java [hunk 7]
```diff
@@ -15,13 +15,13 @@
 public class Value {
 
     public static final String VALUE_CONSTRAINTS =
-            "Cost should only contain numbers and an optional decimal point, which if specified, accepts up to 2"
-            + " decimal digits.";
+            "Value should only contain positive numbers and an optional decimal point, which if specified, "
+                   + "accepts up to 2 decimal digits.\nValue should also be lesser than 1 billion.";
```
Comment (lightz96) [inline src/main/java/thrift/model/transaction/Value.java]: There are only 7 spaces. 😄

File: src/main/java/thrift/storage/JsonAdaptedBudget.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+     * Converts a given {@code Budget} into this class for Jackson use.
+     */
+    public JsonAdaptedBudget(Budget source) {
+        period = source.getBudgetDateString();
+        value = source.getBudgetValue().getUnformattedString();
+    }
+
+    /**
+     * Converts this Jackson-friendly adapted budget object into the model's {@code Budget} object.
+     * @throws IllegalValueException if there are any data constraints violated in the adapted budget.
```
Comment (lightz96) [inline src/main/java/thrift/storage/JsonAdaptedBudget.java:44]: There should be an empty line between description and parameter section.

**Review by lightz96 — CHANGES_REQUESTED** [review]
LGTM. Tested and no bug detected. This PR can be merged after some minor issues have been resolved.

File: docs/DeveloperGuide.adoc (hunk 1 above)
Comment (limerencee) [reply docs/DeveloperGuide.adoc]: Mistake when using the find and replace method to change `c/` to `v/`, thanks for spotting this mistake!

File: src/main/java/thrift/logic/Logic.java (hunk 2 above)
Comment (limerencee) [reply src/main/java/thrift/logic/Logic.java]: Changed to `isRefreshingFilteredList`

File: src/main/java/thrift/logic/parser/BudgetCommandParser.java (hunk 3 above)
Comment (limerencee) [reply src/main/java/thrift/logic/parser/BudgetCommandParser.java:20]: Fixed

File: src/main/java/thrift/model/Model.java (hunk 4 above)
Comment (limerencee) [reply src/main/java/thrift/model/Model.java:149]: Fixed

File: src/main/java/thrift/model/transaction/TransactionIsInMonthYearPredicate.java (hunk 6 above)
Comment (limerencee) [reply src/main/java/thrift/model/transaction/TransactionIsInMonthYearPredicate.java]: Fixed

File: src/main/java/thrift/model/transaction/Value.java (hunk 7 above)
Comment (limerencee) [reply src/main/java/thrift/model/transaction/Value.java]: Fixed

File: src/main/java/thrift/storage/JsonAdaptedBudget.java (hunk 8 above)
Comment (limerencee) [reply src/main/java/thrift/storage/JsonAdaptedBudget.java:44]: Fixed

**Review by lightz96 — APPROVED** [review]

**Pull request merged**